### ISSUE TITLE
Can't reproduce the accuracy of 'EOGHorizontalSignal' dataset in UCR109 based on original 'MINIROCKET' exmples

### ISSUE THREAD

**Issue by firrice** [angus924/minirocket#26]
Hi, as the title suggests, code as follow:

`from minirocket import fit, transform
import pandas as pd
import numpy as np
from sklearn.linear_model import RidgeClassifierCV

csv_root = "/media/data1/ubuntu_env/data/TSC_datasets/Univariate_ts2csv/"
subdataset_name = "EOGHorizontalSignal"
train_csv_file_path = csv_root + subdataset_name + '/{}_TRAIN.csv'.format(subdataset_name)
test_csv_file_path = csv_root + subdataset_name + '/{}_TEST.csv'.format(subdataset_name)

train_csv_file = pd.read_csv(train_csv_file_path,
                        header = None,
                        sep = ",",
                        skiprows = 0,
                        engine = "c")
test_csv_file = pd.read_csv(test_csv_file_path,
                        header = None,
                        sep = ",",
                        skiprows = 0,
                        engine = "c")
total_train_data = train_csv_file.values[:].copy()
X_training, Y_training = total_train_data[:, 1:].astype(np.float32), total_train_data[:, 0].astype(np.int32)
total_test_data = test_csv_file.values[:].copy()
X_test, Y_test = total_test_data[:, 1:].astype(np.float32), total_test_data[:, 0].astype(np.int32)

# note:
# * input time series do *not* need to be normalised
# * input data should be np.float32

parameters = fit(X_training)
X_training_transform = transform(X_training, parameters)
X_test_transform = transform(X_test, parameters)

classifier = RidgeClassifierCV(alphas = np.logspace(-3, 3, 10))
classifier.fit(X_training_transform, Y_training)
predictions = classifier.predict(X_test_transform)
total = len(Y_test)
correct = (predictions == Y_test).sum()
accuracy = correct / total
print("accuarcy: {}".format(accuracy))`

Based on the code above, the final  acc is about 0.59~0.60. But the result to be displayed as follow is 0.83:
![image](https://github.com/angus924/minirocket/assets/26590498/a3538e44-eee8-45c5-8cde-051f4d554079)
I have validated the data loading on several other datasets, accuracy appeared to be normal, so i think data loading in my code should be ok. There is a big difference between 0.58 from 0.80, and i really wounder what's the problem which lead to it. Thanks!

**Comment by angus924** [comment]
Hi @firrice, good question.

The results in `results_ucr109_mean.csv` are mean results, averaged over 30 resamples of the datasets (in doing this we are following common practice in the field, and using the same resamples as in, e.g., https://arxiv.org/abs/2304.13029).

Some of the datasets have strange training/test splits, which means that accuracy can change a lot between the "original" training/test split and once you average over several resamples.

It looks like you are using the original training test splits, so my guess is that is the cause of the difference in accuracy.

On the original training/test splits, I have a result of ~0.5994 for MiniRocket, so more or less what you are getting I think.

I hope that helps.

**Comment by firrice** [comment]
Really thank u for so quick answer! And I will try it on 30 resamples following your suggests :)

**Comment by firrice** [comment]
Hi, I have tried it  on 30 resamples of several datasets in UCR109, and i found accuracy was closed to the result in [https://github.com/angus924/minirocket/blob/main/results/accuracy_ucr109_resamples.csv](url) when normalization was used. I have done a experiment which demonstrated the phenomenon, which could be seen as follow:
![image](https://github.com/angus924/minirocket/assets/26590498/7dbd7f2c-f643-4f72-b980-6d402a10f888)
But in "MiniRocket" of examples, i notice there is:
![image](https://github.com/angus924/minirocket/assets/26590498/fffe1966-85e7-42ce-9334-ef406891cb19)
So i wonder whether u used the normalization in producing the results per resample. 
(PS: I used the 30 resample manner in MultiRocket which refers to [https://arxiv.org/abs/2102.00457](url), and i think it should be same manner in MiniRocket. Or would u mind offering a link to your resample datasets?) Thanks!

**Comment by angus924** [comment]
Hi @firrice, just clarifying, you mean normalization of the input time series (as opposed to normalization of the features produced by minirocket)?

If we're talking about the former, the input time series were not normalized for the minirocket experiments, so I don't think it's that.  More likely the resamples.

The exact samples are a bit of a pain to generate from scratch, as they are generated using specific code / random generation in java in this library: <https://github.com/time-series-machine-learning/tsml-java>.

Indices which allow you to reproduce the resamples for many of the datasets are available here: <https://github.com/ChangWeiTan/MultiRocket/tree/main/data/indices109>.  The way to use these is as follows.  Concatenate the training and test sets, then in the *_INDICES_TRAIN.txt file each row contains the indices for the training set for a given resample (i.e., 30 rows, 30 resamples).  The remaining indices are the test set.  The first row is always all `-1` because the first "resample" is just the original training / test split.

Maybe try with the exact resamples and see what happens.  Could be another issue...

I hope that make sense, let me know!

**Comment by firrice** [comment]
Thanks for your detailed response, as always :) .Firstly, Normalization i refer to means normalization of the input time series, which's form likes: (data - mean) / std. And the resample indices which i have experimented is exactly from [https://github.com/ChangWeiTan/MultiRocket/tree/main/data/indices109](url), and i used the original resample code from the repository.  I have verified that indicies and data loading both have no problem by single step debugging, and ensurly only use the original code.

Here is code which i used:

`from sklearn.linear_model import RidgeClassifierCV
from minirocket import fit, transform
import pandas as pd
import numpy as np
import torch

csv_root = "/media/data1/ubuntu_env/data/TSC_datasets/Univariate_ts2csv/"
resample_indices_root = "/media/data1/ubuntu_env/data/TSC_datasets/UCR109_resample_indices/"
subdataset_name = "EOGHorizontalSignal" #['EOGHorizontalSignal', 'EOGVerticalSignal', 'Ham']
num_classes = 12
resample_id = 1
train_csv_file_path = csv_root + subdataset_name + '/{}_TRAIN.csv'.format(subdataset_name)
test_csv_file_path = csv_root + subdataset_name + '/{}_TEST.csv'.format(subdataset_name)

train_csv_file = pd.read_csv(train_csv_file_path,
                        header = None,
                        sep = ",",
                        skiprows = 0,
                        engine = "c")
test_csv_file = pd.read_csv(test_csv_file_path,
                        header = None,
                        sep = ",",
                        skiprows = 0,
                        engine = "c")
if resample_id == 0:
    total_train_data = train_csv_file.values[:].copy()
    X_training, Y_training = total_train_data[:, 1:].astype(np.float32), total_train_data[:, 0].astype(np.int32)
    total_test_data = test_csv_file.values[:].copy()
    X_test, Y_test = total_test_data[:, 1:].astype(np.float32), total_test_data[:, 0].astype(np.int32)
else:
    train_data = train_csv_file.values[:]
    test_data = test_csv_file.values[:]
    total_data = np.vstack((train_data, test_data))
    all_indices = np.arange(len(total_data))
    train_indices = np.loadtxt(resample_indices_root + "/{}_INDICES_TRAIN.txt".format(subdataset_name),
                                    skiprows=resample_id,
                                    max_rows=1).astype(np.int32)
    test_indices = np.setdiff1d(all_indices, train_indices, assume_unique=True)
    train_resample_data = total_data[train_indices]
    test_resample_data = total_data[test_indices]
    X_training, Y_training = train_resample_data[:, 1:].astype(np.float32), train_resample_data[:, 0].astype(np.int32)
    X_test, Y_test = test_resample_data[:, 1:].astype(np.float32), test_resample_data[:, 0].astype(np.int32)

parameters = fit(X_training)
X_training_transform = transform(X_training, parameters)
X_training_mean = X_training_transform.mean(0)
X_training_std = X_training_transform.std(0) + 1e-8
X_training_transform = (X_training_transform - X_training_mean) / X_training_std # normalise!
X_test_transform = transform(X_test, parameters)
X_training_transform = torch.FloatTensor(X_training_transform)
Y_training = torch.LongTensor(Y_training)
X_test_transform = torch.FloatTensor(X_test_transform)
X_test_transform = (X_test_transform - X_training_mean) / X_training_std # normalise!

classifier = RidgeClassifierCV(alphas = np.logspace(-3, 3, 10))
classifier.fit(X_training_transform, Y_training)
predictions = classifier.predict(X_test_transform)
total = len(Y_test)
correct = (predictions == Y_test).sum()
best_acc = correct / total

print("best accuarcy: {}".format(best_acc))`

For 'EOGHorizontalSignal' dataset which's resample id is 1, using the code above, accuracy was 0.83 when noemalization was used, and accuracy was 0.76 without normalization :(

I will look at it later, maybe it's a little problem that's hidden deep ┓( ´∀` )┏

**Comment by angus924** [comment]
Hi @firrice, I think I see what is happening.

This is an issue about normalizing the output of the minirocket transform, not about normalizing the input time series.  If you notice in your code, you are not doing any modification of the input time series before it is passed to the transform.  You are actually normalizing the output features produced by minirocket.  This is good, this is the correct procedure.

We do want to normalize the features which come out of the transform (as you are doing).

Originally (and I think the code example in the readme is like this), we did this using the "normalize = True" parameter of RidgeClassifierCV.  However, this is now deprecated.

This actually doesn't really have anything do do with minirocket per se, and is instead an issue relating to optimization (in this case the optimization process inside RidgeClassifierCV which learns the parameters of the linear model).

Basically, the ridge classifier (like many other related classifiers or, more accurately, the optimization methods used to learn their parameters), tends to work better if the features are centered and scaled (i.e., given a matrix of features, such that all columns have a mean of approx. zero and a similar standard deviation or some other measure of scale).

The old (now deprecated) "normalize = True" parameter of the ridge classifier used a slightly less common approach of subtracting the mean and dividing by the l2 norm.  To reproduce this, you would do something like this:

```python
# load data
parameters = fit(X_training)
X_training_transform = transform(X_training, parameters)
X_test_transform = transform(X_test, parameters)

_mean = X_training_transform.mean(0)
_norm = np.linalg.norm(X_training_transform - _mean, axis = 0) + 1e-6 # note subtract mean when computing l2

X_training_transform = (X_training_transform - _mean) / _norm
X_test_transform = (X_test_transform - _mean) / _norm
```

Alternatively, you can do what you have done, which is just subtract the mean and divide by the standard deviation.  The results should be similar (I'm pretty sure that subject to mean centering the l2 norm and std are related by a constant factor).

I hope that helps.

**Comment by firrice** [comment]
Really appreciate! @angus924 It's my mistake which i treated the transform in MiniRocket as a kind of preprocess, so mistakenly thought that the normalization was for input.....Thanks again for your patience and detailed answer!
